slim/nets/densenet.py

```diff
@@ -0,0 +1,140 @@
+          is_training: whether is training or not.
+      dense_prediction: Bool, defaults to False
+      reuse: whether or not the network and its variables should be reused. To be
+        able to reuse 'scope' must be given.
+      scope: Optional variable_scope.
+    Returns:
+    net: A rank-4 tensor of size [batch, height_out, width_out, channels_out].
+      if
+    end_points: A dictionary from components of the network to the corresponding
+      activation.
+    """
+    # check n_layers_per_block argument
+    if type(n_layers_per_block) == list:
+        assert (len(n_layers_per_block) == n_dense)
+    elif type(n_layers_per_block) == int:
+        n_layers_per_block = [n_layers_per_block] *  n_dense
+    else:
+        raise ValueError
+
+
+    with tf.variable_scope(scope, 'densenet', [inputs], reuse=reuse) as sc:
+      end_points_collection = sc.name + '_end_points'
+      with slim.arg_scope([slim.conv2d, DenseBlock, TransitionLayer],
+                        outputs_collections=end_points_collection):
+        with slim.arg_scope([slim.batch_norm, slim.dropout],
+                            is_training=is_training):
+
+          #####################
+          # First Convolution #
+          #####################
+          # We perform a first convolution.
+          # If DenseNet BC, first convolution has 2*growth_rate output channels
+          if bottleneck and compression < 1.0:
+            n_filters_first_conv = 2 * growth_rate
+          net = slim.conv2d(inputs, n_filters_first_conv, [7, 7],
+                              stride = [2, 2], scope='first_conv')
+          net = slim.pool(net, [2, 2], stride= [2, 2], pooling_type='MAX')
+          n_filters = n_filters_first_conv
+
+          #####################
+          #    Dense blocks   #
+          #####################
+
+          for i in range(n_dense-1):
+            # Dense Block
+            net, _ = DenseBlock(net, n_layers_per_block[i],
+                                growth_rate, dropout_p,
+                                bottleneck=bottleneck,
+                                scope='denseblock%d' % (i+1))
+            n_filters += n_layers_per_block[i] * growth_rate
+
+            # Transition layer
+            net = TransitionLayer(net, n_filters, dropout_p,
+                                  compression=compression,
+                                  scope='transition%d'%(i+1))
+
+
+          # Final dense block (no transition layer afterwards)
+          net, _ = DenseBlock(net, n_layers_per_block[n_dense-1],
+                              growth_rate, dropout_p,
+                              scope='denseblock%d' % (n_dense))
+
+          #####################
+          #      Outputs      #
+          #####################
+          pool_name = 'pool%d' % (n_dense + 1)
+          if dense_prediction:
+            net = slim.pool(net, [7, 7], pooling_type='AVG', scope=pool_name)
```

Comment:
What if we have not 224x224 input? From my point of view we should use dynamic average pooling based on input shape.